### ISSUE TITLE
Consistency issues with ViewMaximum and ActualMaximum

### ISSUE THREAD

**Issue by stefan-schweiger** [oxyplot/oxyplot#442]
Currently `ActualMaximum` is composed of either `ViewMaximum`, `Maximum` or the *calculated* maximum if the other two are not set. And `ViewMaximum` just gets used when the view gets panned or zoomed (but also has a public setter (why?))

```c#
internal virtual void UpdateActualMaxMin()
{
    if (!double.IsNaN(this.ViewMaximum))
    {
        // The user has zoomed/panned the axis, use the ViewMaximum value.
        this.ActualMaximum = this.ViewMaximum;
    }
    else if (!double.IsNaN(this.Maximum))
    {
        // The Maximum value has been set
        this.ActualMaximum = this.Maximum;
    }
    else
    {
        // Calculate the actual maximum, including padding
        this.ActualMaximum = this.CalculateActualMaximum();
    }

    ...

    this.CoerceActualMaxMin();
}
```

In my opinion `ActualMaximum` should be the actual *calculated* data maximum (+ padding) and `ViewMaximum` the currently *displayed* maximum in the view (and should not have a public setter as it just makes no sense to allow this).

The only two properties with public setters of this kind should be `AbsoluteMaximum` (which should be the maximum value for `ViewMaximum` if set) and `Maximum` (which should be the default value for `ViewMaximum` if set).

```c#
/* This maybe only setting the backing fields in calculation methods
   instead of doing this in the getter is a more suitable way to do this,
   but anyways this is what it could look like.
*/
private double viewMaximum;

public double ViewMaximum
{
    get {
       // if an absolute maximum is set and it's smaller than viewMaximum return AbsoluteMaximum
       if (AbsoluteMaximum < viewMaximum)
           return AbsoluteMaximum;

        // if viewMaximum is not set (no zoom or pan) return Maximum
        return double.IsNaN(viewMaximum) ? Maximum : viewMaximum;
    }
   set { viewMaximum = value; }
}

public double AbsoluteMaximum
{
    get {
        return this.CalculateMaxValue();
    }
}
```

I guess this are just my two cents as it would maybe involve breaking changes for some people (I say maybe because any externally set value for `ViewMaximum` would be replaced pretty fast by the calculated one anyways, the same goes for `AbsoluteMaximum`).

But I want to hear others opinions on this issue, or if there are good reasons for the current implementation I just don't see.

**Comment by objorke** [comment]
I agree that the `ViewMaximum` / `ViewMinimum` setters should not be public.
Changing the implementation of `ActualMaximum` is a breaking change (I am sure many users depend on this one), and needs more consideration...

**Comment by stefan-schweiger** [comment]
Just as a reminder `ActualMaximum` is currently not even accessible in WPF (and all derived implementations), so it's maybe not as many people as you would think.

But it could also be suitable to create a property like `ValueMaximum` to have a property that reflects the calculated maximum to the outside (as there is currently none doing this in a consistent way, and also maybe with this `ActualMaximum` could be marked as obsolete).

What do you think about the changed `ViewMaximum` implementation/calculation?

**Comment by tibel** [comment]
@stefan-schweiger You can use it in WPF when working with the `PlotModel` and `PlotView` (as I do).
Only the `Plot` class that is used to create the plot in XAML has no access to it (XAML only).

**Comment by stefan-schweiger** [comment]
@tibel didn't think about that one. I will take that argument back, but still the current implementation just seems off to me.

**Comment by objorke** [comment]
I don't think we should rename the properties at this point. 

Adding two new properties that contains the global minimum and maximum of all data using the axis sounds OK. 

Do we also need to add local minimum/maximum for the values in the current view?

Ps. I am a bit confused what is the feature actually needed here since the description contains a solution/implementation and not a user story. Is the problem confusing property names, wrong calculation or something else.

**Comment by objorke** [comment]
More thoughts about ViewMaximum/ViewMinimum: 
- it might be ok to keep the setters public as long as it is well described how they work. Maybe someone want to set the initial ViewMinimum/ViewMaximum programatically and also specify Minimum/Maximum at the same time
- or make them completely private/internal if this feature is not really needed

Sorry for the confusion regarding this design - I am not yet sure what is the best solution... 
Need some good arguments if this should be changed...

**Comment by stefan-schweiger** [comment]
I came up with this issues as I was trying to find a solution for https://github.com/oxyplot/oxyplot/issues/424 (which is easier to do with a LockMaximum property than with the binding described in the issue as I always ran in some weird errors, but you still need to know some things about the maximums).

When I was experimenting with the minimums/maximums I just noticed some, at least for me, confusing/not logical things about the implementation of those:

* `ViewMaximum` is not always the maximum of what's currently visible (viewed)
* `ActualMaximum` is closer to what `ViewMaximum`should be
* there is no property to get the calculated maximum once the view was panned or zoomed (as `ActualMaximum` at this point only returns `ViewMaximum`).

**Comment by stefan-schweiger** [comment]
I've looked into some commercial libraries how those handle it (and most only use min and max and that's it) one of those who have properties that are somewhat alike to the oxyplot ones is visifire and I think there naming/implementation is how I think this should be done:

http://www.visifire.com/documentation/Visifire_Documentation/Charts/Reference/Chart_Elements_Reference/Axis.htm#

* **AxisMaximum** Sets the largest value that must appear on the axis.
* **ActualAxisMinimum** Gets the default AxisMinimum which Visifire calculates internally. It is a read-only property.
* **ViewMaximum** Gets axis data view maximum XValue position. It is a read only property.

**Comment by objorke** [comment]
Thanks for the link, I need some time to look into this.

**Comment by stevh177** [comment]
@tibel, I am using PlotView in WPF and I cannot see ActualMaximum or ViewMaximum

**Comment by stefan-schweiger** [comment]
@stevh177 Because there is no WPF Wrapper for those properties implemented.

**Comment by stevh177** [comment]
thanks @stefan-schweiger , seem to be having the same issue with LinearAxis and FormatValue. Are there any plans to add a more extensible wrapper for WPF?

**Comment by stevh177** [comment]
Thanks, @stefan-schweiger. Are there any plans to update the wrapper? I am trying to find a way to get a series of StairStep plots, each a separate control with its own LinearAxis, to be synchronized when the user zooms any one of the plots. Currently, each plot will zoom independently of the others. I am trying to determine if I will need to fork my own implementation, or look to another tool.
